src/middlewared/middlewared/plugins/interface/rdma.py

```diff
@@ -0,0 +1,114 @@
+import json
+import subprocess
+from pathlib import Path
+
+from middlewared.schema import accepts, Dict, List, returns, Str
+from middlewared.service import private, Service
+from middlewared.service_exception import CallError
+from middlewared.utils.functools import cache
+
+PRODUCT_NAME_PREFIX = 'Product Name: '
+SERIAL_NUMBER_PREFIX = '[SN] Serial number: '
+PART_NUMBER_PREFIX = '[PN] Part number: '
+
+
+class InterfaceService(Service):
+
+    class Config:
+        namespace_alias = 'interfaces'
+
+    @private
+    def get_pci_vpd(self, pci_addr):
+        lspci_cmd = ['lspci', '-vv', '-s', pci_addr]
+        ret = subprocess.run(lspci_cmd, capture_output=True)
+        if ret.returncode:
+            self.logger.debug(f'Failed to execute "{" ".join(lspci_cmd)}": {ret.stderr.decode()}')
+            raise CallError(f'Failed to determine serial number/product: {ret.stderr.decode()}')
+        result = {}
+        for line in ret.stdout.decode().split('\n'):
+            sline = line.strip()
+            if sline.startswith(PRODUCT_NAME_PREFIX):
+                result['product'] = sline[len(PRODUCT_NAME_PREFIX):]
+            elif sline.startswith(SERIAL_NUMBER_PREFIX):
+                result['serial'] = sline[len(SERIAL_NUMBER_PREFIX):]
+            elif sline.startswith(PART_NUMBER_PREFIX):
+                result['part'] = sline[len(PART_NUMBER_PREFIX):]
+        return result
+
+    @private
+    @accepts()
+    @returns(List(items=[Dict(
+        'rdma_link_config',
+        Str('rdma', required=True),
+        Str('netdev', required=True),
+    )], register=True))
+    @cache
+    def get_rdma_link_choices(self):
+        """Return a list containing dictionaries with keys 'rdma' and 'netdev'.
+
+        Since these are just the hardware present in the system, we cache the result."""
+        self.logger.info('Fetching RDMA link netdev choices')
+
+        link_cmd = ['rdma', '-j', 'link']
+
+        ret = subprocess.run(link_cmd, capture_output=True)
+        if ret.returncode:
+            self.logger.debug(f'Failed to execute "{" ".join(link_cmd)}": {ret.stderr.decode()}')
+            raise CallError(f'Failed to determine RDMA links: {ret.stderr.decode()}')
+
+        result = []
+        for link in json.loads(ret.stdout.decode()):
+            result.append({'rdma': link['ifname'], 'netdev': link['netdev']})
+        return result
+
+    @accepts()
+    @returns(List(items=[Dict(
+        'rdma_card_config',
+        Str('serial'),
+        Str('product'),
+        Str('part_number'),
+        List('links', items=[Dict('rdma_link_config')])
```

Comment:
Are you wanting to `items=[Ref('rdma_link_config')]`?

Reply:
Yes, thanks.